src/replication.c

```diff
@@ -2885,7 +3688,7 @@ void undoConnectWithPrimary(void) {
  * Never call this function directly, use cancelReplicationHandshake() instead.
  */
 void replicationAbortSyncTransfer(void) {
-    serverAssert(server.repl_state == REPL_STATE_TRANSFER);
+    serverAssert(server.repl_state == REPL_STATE_TRANSFER || server.repl_rdb_conn_state != REPL_RDB_CONN_STATE_NONE);
```

Comment:
this line changes the original intent to allow `repl_state` to be a different value than `REPL_STATE_TRANSFER`. I take it that you meant `&&`? but then in the non-dual-conn case, repl_rdb_conn_state would be NONE, correct?

Reply:
As it stands now, this function deals with the states related to the old style single-connection only. Judging by `cancelReplicationHandshake`, I think you are handling the rdb connection states in `abortRdbConnectionSync` already so `|| server.repl_rdb_conn_state != REPL_RDB_CONN_STATE_NONE`

also I think both names could be improved a bit - `replicationAbortSyncTransfer` and `abortRdbConnectionSync` are actually very related with the former dealing with single-conn full sync while the latter responsible for dual-conn full sync (and disk based ones). In fact, I don't see `replicationAbortSyncTransfer` needs to be invoked for the dual-conn scenario and `abortRdbConnectionSync` is already the superset.

Reply:
`repl_transfer_fd` is used only in the disk based replication. I don't think `abortRdbConnectionSync` needs to worry about it. What I am suggesting 

1. remove the `|| server.repl_rdb_conn_state != REPL_RDB_CONN_STATE_NONE` from the assert in `replicationAbortSyncTransfer` in this PR.
2. `replicationAbortDualConnectionSyncTransfer` looks great. 

We don't need/want to add the `else` in front of `if (server.repl_state == REPL_STATE_TRANSFER)`. The current logic is more cleaner as these are two different states.